lib/posix-timerfd: Validate settime argument

andreittr opened the pull request:
### Description of changes

This change adds a validation check on the `new_value` argument to settime(), refusing to work with negative times and non-canonical timespec values.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

N/A

andreittr committed:
lib/posix-timerfd: Validate settime argument

This change adds a validation check on the `new_value` argument to
settime(), refusing to work with negative times and non-canonical
timespec values.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

mogasergiu commented on lib/posix-timerfd/timerfd.c:
```diff
@@ -314,6 +314,14 @@ int uk_sys_timerfd_settime(const struct uk_file *f, int flags,
 	if (unlikely(f->vol != TIMERFD_VOLID))
 		return -EINVAL;
+	/* Refuse to set negative times, and enforce canonicity */
+	if (unlikely(new_value->it_value.tv_sec < 0 ||
```
Just wondering, would this maybe be better in a macro or static inline instead so that others in this file may possibly use it as well in the future? (also reduces LoC of this function by 20% :sweat_smile:)
Something like `itimerspec_is_valid()`

andreittr replied:
I guess both conditions could be their own macros `TIMESPEC_POSITIVE` and `TIMESPEC_CANON`, and then it'd be
```c
if (!TIMESPEC_POSITIVE(new_value->it_value) ||
    !TIMESPEC_POSITIVE(new_value->it_interval) ||
    !TIMESPEC_CANON(new_value->it_value) ||
    !TIMESPEC_CANON(new_value->it_interval))
        return -EINVAL;
```
with the validation logic maybe packaged in an inline :shrug: .

Mulling it over further, `TIMESPEC_POSITIVE` & `TIMESPEC_CANON` could even be added to `uktimeconv` [timeutil.h](https://github.com/unikraft/unikraft/blob/staging/lib/uktimeconv/include/uk/timeutil.h) as `uk_time_spec_positive` & `uk_time_spec_canon`, respectively.

andreittr committed:
lib/uktimeconv: Add time spec validation inlines

This change adds inlines to check whether a timespec is in canonical
form, and if so whether it represents positive time.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Rebased to current staging

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>
Approved-by: Sergiu Moga <sergiu@unikraft.io>
